runtime/common/src/parachains.rs

```diff
@@ -99,13 +112,65 @@ impl<AccountId, T: Currency<AccountId>> ParachainCurrency<AccountId> for T where
 /// Interface to the persistent (stash) identities of the current validators.
 pub struct ValidatorIdentities<T>(rstd::marker::PhantomData<T>);
 
+#[derive(RuntimeDebug, Encode, Decode)]
+#[derive(Clone, Eq, PartialEq)]
+pub struct DoubleVoteReport {
+	/// Identity of the double-voter.
+	pub identity: ValidatorId,
+	/// First vote of the double-vote.
+	pub first: (Statement, ValidityAttestation),
+	/// Second vote of the double-vote.
+	pub second: (Statement, ValidityAttestation),
+}
+
+impl DoubleVoteReport {
+	fn verify<H: AsRef<[u8]>>(
+		&self,
+		parent_hash: H,
+	) -> Result<(), ()> {
+		let first = self.first.clone();
+		let second = self.second.clone();
+
+		Self::verify_vote(first, &parent_hash, self.identity.clone())?;
```

Comment:
should check that the first statement != the second.

Reply:
Also a test that a double-vote report with both times the same report fails would be good!